data.md

```diff
@@ -795,3 +810,86 @@ Decoding
+      requires ( (START +Int WIDTH) <=Int START0 orBool (START0 +Int WIDTH0) <=Int START )
+
+    // left  margin: [START..(START0..END]..END0)  or  [START..(START0..END0)..END]
+    rule selectRange(storeRange(M, START0, WIDTH0, WS), START, WIDTH) => selectRange(M, START, START0 -Int START)
+                                                                      ++ selectRange(storeRange(M, START0, WIDTH0, WS), START0, (START +Int WIDTH) -Int START0)
+      requires START <Int START0 andBool START0 <Int START +Int WIDTH
+       andBool WIDTH0 >=Int 1 // to avoid unnecessary split
+
+    // right margin: (START0..[START..END0)..END]  or  [START..(START0..END0)..END]
+    rule selectRange(storeRange(M, START0, WIDTH0, WS), START, WIDTH) => selectRange(storeRange(M, START0, WIDTH0, WS), START, (START0 +Int WIDTH0) -Int START)
+                                                                      ++ selectRange(M, START0 +Int WIDTH0, (START +Int WIDTH) -Int (START0 +Int WIDTH0))
+      requires START <Int START0 +Int WIDTH0 andBool START0 +Int WIDTH0 <Int START +Int WIDTH
+       andBool WIDTH0 >=Int 1 // to avoid unnecessary split
+
+    //
+
+    // TODO: check validity: i.e., WS ==K .WordStack
+    rule storeRange(M, _, WIDTH, _) => M requires WIDTH ==Int 0
+
+    rule selectRange(M, _, WIDTH) => .WordStack requires WIDTH ==Int 0
+
+    // lifting
+
+    rule #lookup(M, K)    => select(M, K)   requires notBool (#isConcrete(M) andBool #isConcrete(K))
+    rule #update(M, K, V) => store(M, K, V) requires notBool (#isConcrete(M) andBool #isConcrete(K) andBool #isConcrete(V))
+
+    rule #range(M, START, WIDTH) => selectRange(M, START, WIDTH)                 requires notBool (#isConcrete(M) andBool #isConcrete(START) andBool #isConcrete(WIDTH))
+    rule M [ START := WS ]       => storeRange(M, START, #sizeWordStack(WS), WS) requires notBool (#isConcrete(M) andBool #isConcrete(START) andBool #isConcrete(WS))
+
+    // once it gets down to the concrete map, return to the corresponding function
+    // shouldn't have the infinite rule application
+    rule select(M:Map, K)                 => #lookup(M, K)           [concrete]
+    rule selectRange(M:Map, START, WIDTH) => #range(M, START, WIDTH) [concrete]
+
+    //  syntax IMap ::= ".IMap" [function, smtlib(emptyIMap), smt-prelude] // (define-fun emptyIMap () IMap ((as const IMap) 0))
+//  rule select(.IMap, _) => 0
+
+
+    syntax Bool ::= "#inKeys" "(" IMap "," Int ")" [function]
+
+    rule #inKeys(.IMap, A) => false
+    rule #inKeys(store(M, K, V), K0) => #inKeys(M, K0)
+        requires K =/=Int K0
+
+    rule #inKeys(store(M, K, V), K0) => true
+        requires K ==Int K0
+
+    //Reduces IMaps where multiple entries share the same key
+    rule store(store(M, K0, V0), K1, V1) => store(M, K0, V1)
+        requires K0 ==Int K1
+
+    rule store(store(M, K0, V0), K1, V1) => store(store(M, K1, V1), K0, V0)
+        requires K0 =/=Int K1
+        andBool #inKeys(M, K1)
```

Comment:
^ don't understand this lemma, seems like it might end up getting applied again and again. Are you are trying to change everything into a "normal form" that looks like `store(store(M, K1, V1), K0, V0)` and `K1` is a key of `M`? Do you expect another lemma to be applied immediately afterwards?